pkg/overlay/overlay.go

```diff
@@ -93,6 +106,12 @@ func mountHelper(contentDir, source, dest string, _, _ int, graphOptions []strin
 		// Read-write overlay mounts want a lower, upper and a work layer.
 		workDir := filepath.Join(contentDir, "work")
 		upperDir := filepath.Join(contentDir, "upper")
+
+		if opts != nil {
+			workDir = opts.Workdir
+			upperDir = opts.Upperdir
+		}
+
```

Comment:
@mtrmac Users using custom `upperdir` and `workdir` are already expecting it to be relayed to overlay directly so I think it would better if they get overlay failures.

Reply:
I’m not sure what “directly” means. Is that “a path, with this module responsible for escaping”, “an option on the mount options format, with the caller responsible for (and trusted to do) escaping, if any”, something else?

Reading a _tiny bit_ more, it seems that only the lower directory actually supports a colon-specified list. But right now, the callers seem to be able to pass arbitrary mount options by using e.g. `upperdir=/some/directory,otheroption` . Are the callers known and trusted not to do that, or is that escalating a privilege?

If the callers “should get overlay failures”, I think it would be best to validate the inputs and reject them if they contain unsupported metacharacters. Admittedly the existing `lowerdir` handling does not do any validation either, so this might be fine, assuming all of the data comes from a similarly-trusted source.

Reply:
@mtrmac The value of `upperdir` and `workdir` is user defined so yes if there is vulnerability in overlay fs itself it could cause security issues. But then since its user defined input should we worry about it cause it would be purely end user's fault then. But I am open to suggestions here.

Reply:
If “user” is the human running Podman (and not e.g. contents of a `Containerfile`), that looks fine WRT security, then — it gives the user no more capabilities. Validation still wouldn’t _hurt_, to get a good error message, if nothing else, but might not be essential.


WRT correctness, if it is possible to successfully tell the kernel to use paths that contain `,`, should that be possible using this API? Compare https://unix.stackexchange.com/questions/552630/how-to-escape-comma-on-mount-options-for-overlay . Now there are three ways to go about this (also looking at https://github.com/containers/podman/pull/12712 ):
1. `Workdir` really is a _directory_. The caller can pass `/tmp/a,b`, and it will work, because this code is responsible for escaping. (OTOH if the input in Podman is mount-options-like, then the Podman caller might have to unescape before passing it to this package!)
2. `Workdir` is “a fragment of an overlay mount option”, and the caller is responsible for escaping. (Then the Podman mount-options-like input doesn’t need to be unescaped — _as long as_ Podman is at least separating the individual options correctly)
3. This is too much work to think about, let’s just reject all such input (but we still need to decide what “such input” means; comma? colon? backslash?) for now, and define it later.

Doing nothing now basically commits us to approach 2., because adding unescaping or rejecting some inputs could break callers that start relying on the semantics of 2.

Reply:
Also, if it is 2, the `Workdir` documentation should point out this non-obvious caller responsibility. (I wouldn’t even object to renaming the field to `WorkdirOptionFragment` or something similarly explicit.)

Reply:
@mtrmac I have updated the PR with more inclination towards options `2` with explicit note since I am still unsure what to escape here